### PR TITLE
0.3.0 - Export drivers to userland

### PR DESCRIPTION
**Pull request by lmatteis**
I've separated the two things into separate PRs. This only takes care of putting run in userland without Unified support.

### CODE REVIEW

**Comment by goshakkk** [comment]
Did you think about my idea of scoping the drivers under `cycleMiddleware` returned from `createCycleMiddleware`? We could probably even get rid of `makeXDriver` functions, instead exposing already-made drivers, which makes sense:

```javascript
const cycleMiddleware = createCycleMiddleware();
...

ACTION: cycleMiddleware.actionDriver
```

So that, in the end, `createCycleMiddleware` acts as both:

* middleware constructor
* but also: driver maker, since these are dependent on the middleware anyway

**Comment by lmatteis** [comment]
@goshakkk ok but what would you pass it to `applyMiddleware(...)`? `cycleMiddleware.middleware`? Because `applyMiddleware(...)` wants a function, and in your example `cycleMiddleware` returns an object.

**Comment by goshakkk** [comment]
@lmatteis functions can have properties in js 😈 

```javascript
function createCycleMiddleware() {
  const middleware = () => { ... };
  middleware.actionDriver = ...;
  return middleware;
}

const cycleMiddleware = createCycleMiddleware();
// cycleMiddleware is a fn
cycleMiddleware.actionDriver
// but it also has other fields
```

There's nothing dirty about doing it like that.

Redux-saga does something similar as well: https://github.com/redux-saga/redux-saga#mainjs

**Comment by lmatteis** [comment]
Hrm not really happy about func props. You can't destructor them. I like the idea of having the drivers depend on the middleware

**Comment by goshakkk** [comment]
Fwiw, you can:

```javascript
function a() {}
a.b = 5;
const { b } = a;
console.log(b) ;
// prints 5
```

**Comment by lmatteis** [comment]
@goshakkk ok makes sense. one thing is that i'm still exposing factories (`makeStateDriver` rather than `stateDriver`) because of this:

```js
  cycleMiddleware.makeStateDriver = () => {
    const isSame = {};
    const getCurrent = store.getState;
    return function stateDriver() {
```

Where would `getCurrent` go otherwise (depends on store which is instantiated *after* applyMiddleware is called.

**Comment by goshakkk** [comment]
@lmatteis I think `makeStateDriver` and `makeActionDriver` could be functions inside `createCycleMiddleware`, but what would be exported are not the factory functions themselves, but driver instances:

```js
function createCycleMiddleware() {
  let store;

  function makeStateDriver() { ... }

  const middleware = () => { ... };

  middleware.stateDriver = makeStateDriver();
  return middleware;
}
```

Alternatively, we might not even need the factory functions at all, just assign the driver instances directly:

```js
function createCycleMiddleware() {
  let store;

  function makeStateDriver() { ... }

  const middleware = () => { ... };

  middleware.stateDriver = () => {
    const getCurrent = store.getState;
    return xs.create...;
  };
  return middleware;
}
```

**Comment by lmatteis** [comment]
@goshakkk can we rely on that though? The driver will be executed by Cycle.js, hence `const getCurrent = store.getState;` might get executed in ways we can't predict (multiple times perhaps). Also in terms of API, most cycle drivers are factories. What if we may need to pass arguments in the future?

**Comment by lmatteis** [comment]
@goshakkk i added a `1.0.0` commit with a CHANGELOG.md. Let me know what you think.

**Comment by lmatteis** [comment]
This PR should probably update the code in `example/` with 1.0.0 changes. Problem is that it I can't make it depend on `1.0.0` because it isn't published yet. Should I just import `../` in the examples?

**Comment by lmatteis** [comment]
Another small comment (sorry for the spam): makeStateDriver() needs to be called after run, otherwise store.getState is undefined. Any ideas about this?

**Comment by goshakkk** [comment]
@lmatteis just move that line to be inside `stateDriver`?

File: CHANGELOG.md [hunk 1]
```diff
@@ -0,0 +1,24 @@
+# 1.0.0 - 2017-02-12
+
+**BREAKING CHANGES**
+- `createCycleMiddleware()` no longer takes any arguments. Instead you need to call `Cycle.run` yourself passing it your `main` function and `drivers` explicitly:
```
Comment (goshakkk) [inline CHANGELOG.md]: Is it also worth mentioning here and in the readme the need to `[npm i -s/yarn add] xstream-run`?

**Comment by lmatteis** [comment]
Ok this might be ready to merge. Will wait for the green light from you guys

File: README.md [hunk 2]
```diff
@@ -30,9 +30,10 @@ Handle redux async actions using [Cycle.js](https://cycle.js.org/).
 
 `npm install --save redux-cycles`
 
-Then use `createCycleMiddleware()` which takes as first argument your `main` Cycle.js function, and second argument the Cycle.js drivers you want to use:
+Then use `createCycleMiddleware()` which returns the redux middleware function, but also has two function properties attached to it; namely `makeActionDriver()` and `makeStateDriver()` which you can use accordingly when you call `Cycle.run`.
 
 ```js
+import { run } from '@cycle/xstream-run';
```
Comment (goshakkk) [inline README.md]: Can the readme mention the need to install it as well?

**Comment by lmatteis** [comment]
Well it's a breaking change no? So we need to bump up the first number. 0.3 would mean new features and no breaking changes.

**Comment by goshakkk** [comment]
@lmatteis semver is pretty lax with 0.x's. 0.x means that a project is not yet ready for 1.0, but breaking changes can nevertheless naturally occur — because everything leading up to 1.0 is, in some way, "discovery" of the API, and use cases, and the bugs, and so on; v1.0 signifies the first "stable" / "reliable" version. So it's perfectly fine to make breaking changes and do a 0.3, 0.4, etc. while we're making and testing decisions.

I'd cut 1.0 when we have more people using redux-cycles in production and sharing their feedback, so we will be in position to say 1.0 is truly battle-tested.

More on semver: http://semver.org/#how-should-i-deal-with-revisions-in-the-0yz-initial-development-phase this obviously should be taken as a broad guideline, rather than a strict rule, but I do think this makes sense.

**Comment by lmatteis** [comment]
Right but most users that have `npm i -s redux-cycles` will have `^0.2.3` in their package.json. Meaning that if we release `0.3.0` instead of `1.0.0` it will break their code. I guess we can go for 0.3.0 this since we're not widely used. But the API and code footprint is so small that it's hard to think we'd need to change the API sometimes soon - hence the fast bump to  1.0.0.

**Comment by lmatteis** [comment]
Jeez I didn't know there was so much disagreements with semver: https://github.com/dominictarr/semver-ftw/issues/2#issuecomment-13875242

**Comment by lmatteis** [comment]
Ok I think I took care of all comments. Waiting for your thumbsup @goshakkk

File: example/package.json [hunk 3]
```diff
@@ -22,7 +22,7 @@
     "react-router": "^2.6.1",
     "react-router-redux": "^4.0.5",
     "redux": "^3.5.2",
-    "redux-cycles": "0.2.3",
+    "redux-cycles": "file:../",
```
Comment (goshakkk) [inline example/package.json]: hehe, nice one 👍

**Pull request merged**